### PR TITLE
wip: very rough initial kex traits.

### PR DESCRIPTION
**Pull request by cpu**
## Description

This is a draft PR merging from my ["cpu-djc-crypto++-wip" branch](https://github.com/cpu/rustls/tree/cpu-djc-crypto%2B%2B-wip) into ["cpu-djc-crypto++"](https://github.com/cpu/rustls/tree/cpu-djc-crypto%2B%2B), a branch where I've rebased djc's original crypto backend WIP onto main (and cleaned up a few small things like clippy lints and a failing test).

My intent is that we can iterate on the key exchange specific bits here, and then later use the other branch as a starting point for discussing the larger set of "crypto backend" features.

## Key Exchange Traits

This commit begins the work to abstract key exchange so that non-Ring implementations may be used.

It's very rough! The code builds, but there are no doubt many changes still required :-) In particular I've put little time into thinking through the names of things, or adding Rustdoc commentary. I will also add some self-review comments on parts I think are not quite right or could be improved.

### CODE REVIEW

File: rustls/src/crypto/ring.rs [hunk 1]
```diff
@@ -1,15 +1,28 @@
...
 
 /// Default crypto provider.
 pub struct Ring;
 
+// TODO(XXX): Why do I need to implement this? rustls/tests/api.rs:429 errs otherwise...
+impl Debug for Ring {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str("*ring*")
+    }
+}
```
Comment (cpu) [inline rustls/src/crypto/ring.rs:21]: As the comment mentions, the compiler forced me into implementing this trait for the `Ring` type but I'm not sure _why_. This impl is obviously not very useful and so there's probably a better way to fix the compile err I'm missing.

File: rustls/src/lib.rs [hunk 2]
```diff
@@ -260,12 +260,12 @@
 // Require docs for public APIs, deny unsafe code, etc.
 #![forbid(unsafe_code, unused_must_use)]
 #![cfg_attr(not(read_buf), forbid(unstable_features))]
+// TODO(@cpu): Restore the doc requirement clippy!!!
```
Comment (cpu) [inline rustls/src/lib.rs:263]: I turned this off to avoid needing to write rustdoc for code that's still in flux. I promise to restore it :-)

File: rustls/src/lib.rs [hunk 3]
```diff
@@ -490,15 +488,6 @@ pub mod version {
...
 }
 
-/// All defined key exchange groups appear in this module.
-///
-/// ALL_KX_GROUPS is provided as an array of all of these values.
-pub mod kx_group {
-    pub use crate::kx::SECP256R1;
-    pub use crate::kx::SECP384R1;
-    pub use crate::kx::X25519;
-}
```
Comment (cpu) [inline rustls/src/lib.rs:500]: I'm not sure I fully understand why this module exists, so I've removed it for now. It might be something that should stay removed now that which KX groups are supported is a property of a specific crypto backend.

File: rustls/src/client/client_conn.rs [hunk 4]
```diff
@@ -130,7 +129,7 @@ pub trait ResolvesClientCert: Send + Sync {
 /// * [`ClientConfig::session_storage`]: the default stores 256 sessions in memory.
 /// * [`ClientConfig::alpn_protocols`]: the default is empty -- no ALPN protocol is negotiated.
 /// * [`ClientConfig::key_log`]: key material is not logged.
-pub struct ClientConfig<C> {
```
Comment (cpu) [inline rustls/src/client/client_conn.rs:133]: I had to tighten this bound from `C` to `C: CryptoProvider` in order to have the `kx_groups` field rely on the associated `KeyExchange` type (and in turn, its associated `SupportedGroup` type). Changing this bound cascaded to a few other places (e.g. the `clone` impl). The same flavour of change happened in a couple other places.

File: rustls/src/client/tls12.rs [hunk 5]
```diff
@@ -1063,19 +1062,21 @@ impl<C: CryptoProvider> State<ClientConnectionData> for ExpectFinished<C> {
...
 
 // -- Traffic transit state --
-struct ExpectTraffic {
-    secrets: ConnectionSecrets,
+struct ExpectTraffic<C: CryptoProvider> {
+    secrets: ConnectionSecrets<C>,
     _cert_verified: verify::ServerCertVerified,
     _sig_verified: verify::HandshakeSignatureValid,
     _fin_verified: verify::FinishedMessageVerified,
+    backend: PhantomData<C>,
```
Comment (cpu) [inline rustls/src/client/tls12.rs:1076]: The `ConnectionSecrets` type is newly generic since it relies on `C::KeyExchange`. Since it didn't directly use `C` I had to add a PhantomData<C> here and in the `ConnectionSecrets` struct. This feels a little odd (also the field should probably be named `_backend`?).

Maybe the solution should be to make some functions of `ConnectionSecrets` generic, but not the struct? :thinking:

File: rustls/src/crypto.rs [hunk 6]
```diff
@@ -1,11 +1,17 @@
...
 use crate::server::ProducesTickets;
+use crate::{Error, NamedGroup};
 
 /// *ring* based CryptoProvider.
 pub mod ring;
 
 /// Pluggable crypto galore.
 pub trait CryptoProvider: Send + Sync + 'static {
+    /// KeyExchange operations that are supported by the provider.
+    type KeyExchange: KeyExchange;
```
Comment (cpu) [inline rustls/src/crypto.rs:13]: Probably a place where the associated type name shouldn't equal the trait constraint name.

File: rustls/src/crypto.rs [hunk 7]
```diff
@@ -15,3 +21,34 @@ pub trait CryptoProvider: Send + Sync + 'static {
...
 }
+
+pub enum KeyExchangeError {
+    UnsupportedGroup,
+    KeyExchangeFailed(GetRandomFailed),
+}
+
+/// KeyExchange supports performing a key exchange with a peer using a supported group.
+pub trait KeyExchange: Send + Sync + 'static {
+    type SupportedGroup: SupportedGroup;
```
Comment (cpu) [inline rustls/src/crypto.rs:32]: ditto here RE: associated type name matching the trait constraint name.

File: rustls/src/tls12/mod.rs [hunk 8]
```diff
@@ -198,15 +199,16 @@ impl fmt::Debug for Tls12CipherSuite {
...
-pub(crate) struct ConnectionSecrets {
+pub(crate) struct ConnectionSecrets<C: CryptoProvider> {
     pub(crate) randoms: ConnectionRandoms,
     suite: &'static Tls12CipherSuite,
     pub(crate) master_secret: [u8; 48],
+    backend: PhantomData<C>,
 }
 
-impl ConnectionSecrets {
+impl<C: CryptoProvider> ConnectionSecrets<C> {
```
Comment (cpu) [inline rustls/src/tls12/mod.rs:209]: Here I've made this `impl` block generic over `CryptoProvider` **but** several of the impl functions are in-fact reliant on `Ring` in particular. E.g. `make_cipher_pair` takes a `aead::Algorithm` parameter.

I initially had two `impl` blocks, one that was `<C: CryptoProvider>` and one that was `<Ring>`, but then I had compiler errors from call-sites that were `<C: CryptoProvider>` constrained but needed to call these functions.

My thinking was we'd want to include these functions in the `<C: CryptoProvider>` block until we have more traits that we can use to pull replace the Ring specific bits. This may/may not match what you had in mind!

File: rustls/src/server/tls12.rs [hunk 9]
```diff
@@ -393,16 +395,16 @@ mod client_hello {
         common.send_msg(c, false);
     }
 
-    fn emit_server_kx(
+    fn emit_server_kx<C: CryptoProvider>(
         transcript: &mut HandshakeHash,
         common: &mut CommonState,
         sigschemes: Vec<SignatureScheme>,
-        skxg: &'static kx::SupportedKxGroup,
+        skxg: &'static <<C as CryptoProvider>::KeyExchange as KeyExchange>::SupportedGroup,
```
Comment (cpu) [inline rustls/src/server/tls12.rs:402]: There are a number of places where the compiler forced me to disambiguate the traits but this is probably the grossest since we have to disambiguate the CryptoProvider to get the KeyExchange, and then the KeyExchange to get the SupportedGroup. 

It feels odd to me that there's ambiguity here but maybe this is common-place in this sort of Rust code?

**Comment by cpu** [comment]
@djc This is the draft PR we discussed briefly in Discord DMs. Thanks!

File: rustls/src/crypto/ring.rs (hunk 1 above)
Comment (djc) [reply rustls/src/crypto/ring.rs:21]: I guess we're somehow leaking the `CryptoProvider` type into a `Debug` impl of some other type? I'm looking at the `impl fmt::Debug for ConfigBuilder` on my original crypto branch but it doesn't seem to impose a `Debug` bound on the `CryptoProvider` trait... You'd have to dig into that error message more deeply.

File: rustls/src/lib.rs (hunk 2 above)
Comment (djc) [reply rustls/src/lib.rs:263]: Sounds about right! :)

File: rustls/src/lib.rs (hunk 3 above)
Comment (djc) [reply rustls/src/lib.rs:500]: The `ConfigBuilder` allows the caller to supply an explicit set of key exchange groups. This is part of a scheme to allow applications to select at compile-time which algorithms are used, and makes sure that the unused algorithms don't get compiled into the binary... but as part of that you have to actually supply the symbols that are linked to those algorithms somehow.

File: rustls/src/client/client_conn.rs (hunk 4 above)
Comment (djc) [reply rustls/src/client/client_conn.rs:133]: Yup, that isn't really avoidable and should be mostly fine.

File: rustls/examples/internal/bogo_shim.rs [hunk 10]
```diff
@@ -323,11 +323,11 @@ fn lookup_scheme(scheme: u16) -> rustls::SignatureScheme {
     }
 }
 
-fn lookup_kx_group(group: u16) -> &'static rustls::SupportedKxGroup {
+fn lookup_kx_group(group: u16) -> &'static rustls::crypto::ring::SupportedKxGroup {
     match group {
-        0x001d => &rustls::kx_group::X25519,
-        0x0017 => &rustls::kx_group::SECP256R1,
-        0x0018 => &rustls::kx_group::SECP384R1,
+        0x001d => &rustls::crypto::ring::X25519,
```
Comment (djc) [inline rustls/examples/internal/bogo_shim.rs:328]: I think we should keep the top-level public API unchanged here for a bit, then we can remodel it at the end.

File: rustls/examples/internal/bogo_shim.rs [hunk 11]
```diff
@@ -414,7 +414,7 @@ fn make_server_cfg(opts: &Options) -> Arc<rustls::ServerConfig<Ring>> {
             .map(|curveid| lookup_kx_group(*curveid))
             .collect()
     } else {
-        rustls::ALL_KX_GROUPS.to_vec()
+        rustls::crypto::ring::RingKeyExchange::all_supported_groups().to_vec()
```
Comment (djc) [inline rustls/examples/internal/bogo_shim.rs:417]: Better to just import `RingKeyExchange` somewhere. I guess we have to use a method because we can't have an associated `const` (associated `static`s aren't a thing)? I think I'd call it `all_kx_groups()`, the "supported" aspect kind of flows from the types.

File: rustls/examples/internal/bogo_shim.rs [hunk 12]
```diff
@@ -534,7 +534,7 @@ fn make_client_cfg(opts: &Options) -> Arc<rustls::ClientConfig<Ring>> {
...
```
Comment (djc) [inline rustls/examples/internal/bogo_shim.rs:537]: This gets a bit wordy, just import `RingKeyExchange`?

File: rustls/src/client/builder.rs [hunk 13]
```diff
@@ -1,26 +1,25 @@
+use std::marker::PhantomData;
+use std::sync::Arc;
+use std::time::SystemTime;
```
Comment (djc) [inline rustls/src/client/builder.rs:3]: Although I'm not a big fan, prevailing style is that import blocks are in most to least-specific order, so `std` imports come in the last (unlike in most other crates). I've mostly tried to be consistent with that style.

File: rustls/src/client/hs.rs [hunk 14]
```diff
@@ -1,12 +1,16 @@
+use std::sync::Arc;
+
 #[cfg(feature = "logging")]
 use crate::bs_debug;
 use crate::check::inappropriate_handshake_message;
+use crate::client::client_conn::ClientConnectionData;
+use crate::client::common::ClientHelloDetails;
+use crate::client::{tls13, ClientConfig, ServerName};
```
Comment (djc) [inline rustls/src/client/hs.rs:8]: I usually prefer to avoid unrelated changes, makes it easier to review.

File: rustls/src/client/hs.rs [hunk 15]
```diff
@@ -141,7 +140,7 @@ pub(super) fn start_handshake<C: CryptoProvider>(
     let hello_details = ClientHelloDetails::new();
     let sent_tls13_fake_ccs = false;
     let may_send_sct_list = config.verifier.request_scts();
-    Ok(emit_client_hello_for_retry(
+    Ok(emit_client_hello_for_retry::<C>(
```
Comment (djc) [inline rustls/src/client/hs.rs:143]: Is this really necessary? Why? I would assume that the parameter gets inferred from the function arguments we pass in.

File: rustls/src/client/hs.rs [hunk 16]
```diff
@@ -752,7 +751,7 @@ impl<C: CryptoProvider> ExpectServerHelloOrHelloRetryRequest<C> {
 
         let key_share = match req_group {
             Some(group) if group != offered_key_share.group() => {
-                match kx::KeyExchange::choose(group, &self.next.config.kx_groups) {
+                match KeyExchange::choose(group, self.next.config.kx_groups.as_slice()) {
```
Comment (djc) [inline rustls/src/client/hs.rs:754]: Usually `&Vec<_>` will coerce to the slice, is that not the case here?

File: rustls/src/client/tls12.rs [hunk 17]
```diff
@@ -453,9 +451,9 @@ fn emit_certificate(
     common.send_msg(cert, false);
 }
 
-fn emit_clientkx(transcript: &mut HandshakeHash, common: &mut CommonState, pubkey: &PublicKey) {
+fn emit_clientkx(transcript: &mut HandshakeHash, common: &mut CommonState, pubkey: &[u8]) {
```
Comment (djc) [inline rustls/src/client/tls12.rs:454]: This might make a good candidate for a separate commit.

File: rustls/src/tls12/mod.rs [hunk 18]
```diff
@@ -198,15 +199,16 @@ impl fmt::Debug for Tls12CipherSuite {
 }
 
 /// TLS1.2 per-connection keying material
-pub(crate) struct ConnectionSecrets {
+pub(crate) struct ConnectionSecrets<C: CryptoProvider> {
```
Comment (djc) [inline rustls/src/tls12/mod.rs:202]: Eventually this type will probably need a `CryptoProvider` argument for its `suite` field, but for now I'd just keep it out and pass it to methods that need it. When we do need a `PhantomData`, I'd probably call  it `provider` to match the trait name, though `_provider` also seems fine. I suppose my philosophy is you don't really need the `_` prefix because the `PhantomData` type is right there and there's nothing you can do with it, anyway.

File: rustls/src/tls12/mod.rs (hunk 8 above)
Comment (djc) [reply rustls/src/tls12/mod.rs:209]: Yeah, I think it makes sense to rely on *ring* code directly for now. Maybe this becomes the next part to try to generalize?

File: rustls/src/tls12/mod.rs [hunk 19]
```diff
@@ -240,6 +243,40 @@ impl ConnectionSecrets {
         Ok(ret)
     }
 
+    pub(crate) fn suite(&self) -> &'static Tls12CipherSuite {
```
Comment (djc) [inline rustls/src/tls12/mod.rs:246]: Don't move this stuff around in the same commit, please, makes it harder to see if you made any other changes to it.

File: rustls/src/crypto.rs (hunk 6 above)
Comment (djc) [reply rustls/src/crypto.rs:13]: Hmm, I kinda like it this way.

File: rustls/src/crypto.rs [hunk 20]
```diff
@@ -15,3 +21,34 @@ pub trait CryptoProvider: Send + Sync + 'static {
...
+    fn choose(
+        name: NamedGroup,
+        supported: &[&'static Self::SupportedGroup],
+    ) -> Result<Self, KeyExchangeError>
+    where
+        Self: Sized;
+
+    fn start(skxg: &'static Self::SupportedGroup) -> Result<Self, GetRandomFailed>
+    where
+        Self: Sized;
```
Comment (djc) [inline rustls/src/crypto.rs:45]: I think we can just put the `Sized` bound on the trait itself (with `Send + Sync + 'static`).

File: rustls/src/crypto.rs [hunk 21]
```diff
@@ -15,3 +21,34 @@ pub trait CryptoProvider: Send + Sync + 'static {
...
+
+    fn group(&self) -> NamedGroup;
+    fn pubkey(&self) -> &[u8];
```
Comment (djc) [inline rustls/src/crypto.rs:48]: `pubkey` is actually an unidiomatic name, let's take the opportunity to use `pub_key` instead.

https://rust-lang.github.io/api-guidelines/naming.html

File: rustls/src/crypto.rs [hunk 22]
```diff
@@ -15,3 +21,34 @@ pub trait CryptoProvider: Send + Sync + 'static {
...
+    fn complete<T>(self, peer: &[u8], f: impl FnOnce(&[u8]) -> Result<T, ()>) -> Result<T, Error>;
+}
+
+pub trait SupportedGroup: Clone + Debug + Send + Sync + 'static {
```
Comment (djc) [inline rustls/src/crypto.rs:52]: Why does this need `Clone`? I don't think that should be necessary.

File: rustls/src/crypto/ring.rs [hunk 23]
```diff
@@ -95,3 +108,123 @@ impl ProducesTickets for AeadTicketer {
         Some(out)
     }
 }
+
+/// An in-progress key exchange. This has the algorithm,
+/// our private key, and our public key.
+pub struct RingKeyExchange {
```
Comment (djc) [inline rustls/src/crypto/ring.rs:114]: I don't think we should call these types `Ring*`. Instead, we can refer to them as `ring::KeyExchange`, avoiding what is sometimes referred to as "stuttering". Then we can refer to the traits as `crypto::KeyExchange` in this module.

Since this is largely the same type as before, I'd try to move it into its new place in a separate commit, so that the changes to implement the trait become more obvious when reviewing this commit.

File: rustls/src/crypto/ring.rs [hunk 24]
```diff
@@ -95,3 +108,123 @@ impl ProducesTickets for AeadTicketer {
...
+impl Debug for RingKeyExchange {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.skxg.fmt(f)
+    }
+}
+
+impl KeyExchange for RingKeyExchange {
+    type SupportedGroup = SupportedKxGroup;
+
+    fn all_supported_groups() -> &'static [&'static Self::SupportedGroup] {
```
Comment (djc) [inline rustls/src/crypto/ring.rs:129]: I like to lay out methods in a sort of top-down order. Constructors come first, trivial functions come last. More complex before least complex. Dependent before dependee. `&mut self` receivers before `&self` receivers.

File: rustls/src/server/tls12.rs (hunk 9 above)
Comment (djc) [reply rustls/src/server/tls12.rs:402]: The type `C` could also implement another trait that also has a `KeyExchange` associated type, so Rust makes you spell it out to make sure to get the right one. It's verbose, but not much we can do about it.

File: rustls/src/server/tls12.rs [hunk 25]
```diff
@@ -393,16 +395,16 @@ mod client_hello {
...
         signing_key: &dyn sign::SigningKey,
         randoms: &ConnectionRandoms,
-    ) -> Result<kx::KeyExchange, Error> {
-        let kx = kx::KeyExchange::start(skxg)?;
-        let secdh = ServerECDHParams::new(skxg.name, kx.pubkey.as_ref());
+    ) -> Result<C::KeyExchange, Error> {
+        let kx: <C as CryptoProvider>::KeyExchange = KeyExchange::start(skxg)?;
```
Comment (djc) [inline rustls/src/server/tls12.rs:406]: I usually prefer to avoid type annotations everywhere. If necessary, I prefer to turbofish them (`KeyExchange::<..>::start(skxg)`). But, I think it probably shouldn't be necessary here?

**Comment by djc** [comment]
I think my preferred way to review this would be to look at the PR for the entire branch at once? Then I can navigate to earlier commits, making it easier to refer to earlier changes you've made (as they started to deviate from my original branch).

Making this branch easy to review is a high priority IMO, since it'll be large and slightly longer-lived, I'll review it multiple times in its entirety, and we'll have to get it reviewed by ctz at some point too.

**Comment by cpu** [comment]
> I think my preferred way to review this would be to look at the PR for the entire branch at once

Closing this PR. I'll open a fresh one matching this preference shortly.

We've chatted OOB but saying it again: Thanks for the review!
